### PR TITLE
Add cluster setup for testing CRI-O on GCE

### PR DESCRIPTION
**Pull request by mrunalp**
Add profile for creating CRI-O cluster for CI on GCE.

@runcom @lsm5 
Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

### CODE REVIEW

**Comment by mrunalp** [comment]
@lsm5 dockertested repo configured here doesn't have the cri-o rpm. Could you help with that?

**Comment by runcom** [comment]
We need to import cri-o and cri-tools in dockertested

**Comment by lsm5** [comment]
@runcom @mrunalp how comfortable are we with switching to CentOS Virt SIG/PaaS SIG repos? For example for origin 3.9 we have https://buildlogs.centos.org/centos/7/paas/x86_64/openshift-origin39/ . dockertested just seems like a burden that nobody wants to care about. So CentOS repos would be a really cheap alternative.

**Comment by lsm5** [comment]
my plan RE: PaaS SIG repos is to have -testing repos on test nodes, and once all tests pass, we could move the cri-o/cri-tools packages to -release repos. Production nodes use -release repos. Same with our other packages.

File: cluster/test-deploy/gcp-crio/vars-origin.yaml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+openshift_hosted_registry_storage_kind: object
+openshift_hosted_registry_storage_provider: gcs
+openshift_hosted_registry_storage_gcs_bucket: origin-ci-test-{{ lookup('env', 'INSTANCE_PREFIX') | mandatory }}
+
+container_runtime_docker_storage_type: overlay2
+
+provision_custom_repositories: [
+  {
+    "name": "dockertested",
+    "baseurl": "https://mirror.ops.rhcloud.com/enterprise/rhel/dockertested/x86_64/os/
```
Comment (mrunalp) [inline cluster/test-deploy/gcp-crio/vars-origin.yaml]: @smarterclayton who has access to add rpms to this repo?
cc: @lsm5

Comment (lsm5) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: @mrunalp so the thing with dockertested is it's not cri-o version specific. So having cri-o 1.9, 1.10, 1.11 in the same repo is sure to cause problems. We'll need additional version-specific yum repos specifically to handle cri-o and cri-tools, which is what the CentOS PaaS SIG provides already.

Comment (lsm5) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: and as for docker and other packages which aren't openshift version specific, I'm hoping we can use CentOS Virt SIG, so there would be no need to keep dockertested moving forward.

Comment (lsm5) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: I gotta get CentOS rel-eng to set them up.

Comment (lsm5) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: @mrunalp RE: https://cbs.centos.org/repos/paas7-openshift-origin310-candidate/x86_64/os/ there's no ssl or gpg checks to do, so a simple: sslverify: false and gpgcheck: 0 should suffice along with the repo link. Does that help?

Comment (lsm5) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: let me add cri-tools in there and get back..

Comment (mrunalp) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: @lsm5 thanks, I'll try it out once you add cri-tools.

File: cluster/test-deploy/gcp-crio/vars-origin.yaml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        https://euw-mirror1.ops.rhcloud.com/enterprise/rhel/dockertested/x86_64/os/
+        https://gce-mirror1.ops.rhcloud.com/enterprise/rhel/dockertested/x86_64/os/
+        https://gce-mirror2.ops.rhcloud.com/enterprise/rhel/dockertested/x86_64/os/",
+    "sslverify": "false",
+    "sslclientkey": "ops-mirror.pem",
+    "sslclientcert": "ops-mirror.pem",
+    "failovermethod": "priority",
+    "gpgcheck": 0,
+    "gpgkey": "https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-release https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-beta https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-openshifthosted"
+  }
```
Comment (lsm5) [inline cluster/test-deploy/gcp-crio/vars-origin.yaml]: ok so the cbs.centos.org repo mentioned above has cri-tools now. Can you add the following and check how it goes..


{
"name": "CentOS PaaS SIG - Origin 3.10",
"baseurl": https://cbs.centos.org/repos/paas7-openshift-origin310-candidate/x86_64/os/",
"sslverify": "false",
"gpgcheck": 0
}

Comment (lsm5) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: @mrunalp ^

Comment (mrunalp) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: Yeah, trying it out. Thanks!

File: cluster/test-deploy/gcp-crio/vars-origin.yaml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    "failovermethod": "priority",
+    "gpgcheck": 0,
+    "gpgkey": "https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-release https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-beta https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-openshifthosted"
+  },
+  {
+    "name": "CentOS PaaS SIG - Origin 3.10",
+    "baseurl": "https://cbs.centos.org/repos/paas7-openshift-origin310-candidate/x86_64/os/",
+    "sslverify": "false",
+    "gpgcheck": 0
+  }
```
Comment (lsm5) [inline cluster/test-deploy/gcp-crio/vars-origin.yaml]: Try this.. I think apart from name and and baseurl, all parameters should be the same as dockertested.

{
    "name": "cri-o-tested 3.10",
    "baseurl": "https://mirror.openshift.com/enterprise/rhel/cri-o-tested/3.10/x86_64/os",
    "sslverify": "false",
    "sslclientkey": "ops-mirror.pem",
    "sslclientcert": "ops-mirror.pem",
    "failovermethod": "priority",
    "gpgcheck": 0,
    "gpgkey": "https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-release https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-beta https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-openshifthosted"
  },

Comment (lsm5) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: @mrunalp ^

Comment (mrunalp) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: Trying that now.

File: cluster/test-deploy/gcp-crio/vars-origin.yaml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    "sslclientkey": "ops-mirror.pem",
+    "sslclientcert": "ops-mirror.pem",
+    "failovermethod": "priority",
+    "gpgcheck": 0,
+    "gpgkey": "https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-release https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-beta https://mirror.ops.rhcloud.com/libra/keys/RPM-GPG-KEY-redhat-openshifthosted"
+  },
+  {
+    "name": "CentOS PaaS SIG - Origin 3.10",
+    "baseurl": "https://cbs.centos.org/repos/paas7-openshift-origin310-candidate/x86_64/os/",
+    "sslverify": "false",
```
Comment (lsm5) [inline cluster/test-deploy/gcp-crio/vars-origin.yaml]: "enabled": 1,

@mrunalp ^

**Comment by mrunalp** [comment]
The PR in its current form works but still uses the centos paas repo. @lsm5 and I will look to switching to cri-o-tested instead. Also, we still need to figure out handling different versions for PRs targeted towards different release branches.

**Comment by lsm5** [comment]
On Tue, Jul 03, 2018 at 03:47:42PM -0700, Mrunal Patel wrote:
> The PR in its current form works but still uses the centos paas repo. @lsm5 and I will look to switching to cri-o-tested instead. Also, we still need to figure out handling different versions for PRs targeted towards different release branches. 
> 

I just need to create the additional dirs for other versions and generate
repodata, and then we'll be all set. I'll get this done tomorrow.

RE: cri-o-tested I'll check with openshift
people about pem files.

-- 
Lokesh
IRC: lsm5
GPG: 0xC7C3A0DD
https://keybase.io/lsm5

**Comment by cevich** [comment]
> Also, we still need to figure out handling different versions for PRs targeted towards different release branches

Target-branch information and commit range, is surprisingly difficult to "discover" just based on a repository clone.  Therefor, it's typically exposed via an environment variable by the CI system itself (which obtains it from github metadata).  I have yet to work on a CI system that doesn't make it available.  The solution normally boils down to some sort of lookup table.  For example,

```
case $BRANCH
   1.0*) VAR1=foo; VAR2=bar; ...;;
   1.1*) VAR1=sna; VAR2=foo; ...;;
...snip...
do_stuffs.sh $VAR1 $VAR2
```

In this way ``do_stuffs.sh`` (example script that runs the tests) has the details to put other components together correctly.  For example, CRI-O 1.10 + OpenShift 3.10.

The real challenge is in how to also handle testing of multiple distros, and/or substantial configuration variations.  Any testing matrix with more than one dimension, compounds in scale and complexity immensely.   Every CI system handles matrices differently, and this greatly impacts the maintenance overhead (and frustration).

The most naive/simple method is to enumerate the entire matrix as individual jobs.  Then somehow skip jobs which are unsupported/impossible/incompatible, or rig their results as an automatic **PASS**.  This solution typically has the highest maintenance burden.

More sophisticated CI systems allow jobs to trigger other jobs, and aggregate the results.  Typically the key, or label for each result is taken as some kind of hash of the matrix variables.  That could be literal concatenation, an actual hash, or some combination.  This solution is (obviously) a lot more complex, but pays back over the long-term by reduced maintenance/upkeep costs.

Hope that helps.

**Comment by lsm5** [comment]
@mrunalp 
cri-o 1.11 https://cbs.centos.org/repos/paas7-openshift-origin311-candidate/x86_64/os/ 
and cri-o 1.9 https://cbs.centos.org/repos/paas7-openshift-origin39-candidate/x86_64/os should be ready to go. Let me know if any issues..

**Comment by runcom** [comment]
Works perfectly, got a crio cluster up and running

**Comment by cevich** [comment]
@mrunalp what r your thoughts WRT this, vs the openshift-ansible installer PR tester in origin (I'm assuming this PR is related).  In any case, let me know if you'd like me to (and/or help ) write origin-ci jobs for this.

Assuming so, do note: Even with origin-ci jobs, I still don't know how/why/what mechanism actually posts the results back to PRs.  So quite likely we'd be in the same boat.  So let me know how you'd like to proceed.

Related: I posted an e-mail to the group regarding alternate CI systems, and a cache-image management tool.  Both support GCE, and in-repo definitions for their respective pieces.

**Comment by lsm5** [comment]
@mrunalp cri-o-tested should be ready for 3.9 and 3.11 as well. Added the corresponding cri-o and cri-tools packages there.

**Comment by mrunalp** [comment]
@lsm5 Thanks! Testing it out.

**Comment by lsm5** [comment]
@mrunalp i now have `latest -> 3.11` for cri-o-tested.

File: cluster/test-deploy/gcp-crio/vars-origin.yaml [hunk 5]
```diff
@@ -0,0 +1,108 @@
+# Package and target versions
+ 
+openshift_use_crio: true
```
Comment (smarterclayton) [inline cluster/test-deploy/gcp-crio/vars-origin.yaml]: This is the file that is supposed to be the same across all variants.  You're only supposed to modify vars.yaml.

**Comment by smarterclayton** [comment]
Another note - crio is going to be part of the openshift channels.  Whatever structure is being built here for those packages needs to be aligned with that (i.e. when we do official build channels we're going to use the crio that comes out of the brew tag).

File: cluster/test-deploy/gcp-crio/vars-origin.yaml (hunk 5 above)
Comment (mrunalp) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: okay, I will move this to vars.yml.

Comment (mrunalp) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: What about the repo that we added to openshift_additional_repos?

Comment (mrunalp) [reply cluster/test-deploy/gcp-crio/vars-origin.yaml]: It might be getting merged. I'll try it out and read docs.

**Comment by mrunalp** [comment]
Pushed changes to move customizations to vars.yml

File: cluster/test-deploy/gcp-crio/vars.yaml [hunk 6]
```diff
@@ -1,4 +1,36 @@
 ---
+provision_custom_repositories: [
```
Comment (runcom) [inline cluster/test-deploy/gcp-crio/vars.yaml]: @smarterclayton for some reason, when we provision a new cluster (`make WHAT ....`), these repositories here are entirely skipped as if the `vars.yaml` file isn't taken into account _at all_. Cluster setup fails then as no dockertested/criotested repo will be on the system.

**Comment by mrunalp** [comment]
This works with @runcom ansible PR.

**Comment by mrunalp** [comment]
@smarterclayton I have updated provision_custom_repositories for all profiles as discussed.

**Pull request merged**